src/gc.c

```diff
@@ -1986,34 +2043,38 @@ STATIC_INLINE void gc_try_claim_and_push(jl_gc_markqueue_t *mq, void *_obj,
 }
 
 // Mark object with 8bit field descriptors
-STATIC_INLINE jl_value_t *gc_mark_obj8(jl_ptls_t ptls, char *obj8_parent, uint8_t *obj8_begin,
+STATIC_INLINE jl_value_t* gc_mark_obj8(jl_ptls_t ptls, char *obj8_parent, uint8_t *obj8_begin,
                          uint8_t *obj8_end, uintptr_t nptr) JL_NOTSAFEPOINT
 {
     (void)jl_assume(obj8_begin < obj8_end);
     jl_gc_markqueue_t *mq = &ptls->mark_queue;
     jl_value_t **slot = NULL;
     jl_value_t *new_obj = NULL;
+    int push_into_queue = pb_queue_free_space(mq) > 33;
     for (; obj8_begin < obj8_end; obj8_begin++) {
-        slot = &((jl_value_t**)obj8_parent)[*obj8_begin];
-        new_obj = *slot;
+        new_obj = ((jl_value_t**)obj8_parent)[*obj8_begin];
         if (new_obj != NULL) {
             verify_parent2("object", obj8_parent, slot, "field(%d)",
                             gc_slot_to_fieldidx(obj8_parent, slot, (jl_datatype_t*)jl_typeof(obj8_parent)));
-            if (obj8_begin + 1 != obj8_end) {
-                gc_try_claim_and_push(mq, new_obj, &nptr);
-            }
-            else {
-                // Unroll marking of last item to avoid pushing
-                // and popping it right away
-                jl_taggedvalue_t *o = jl_astaggedvalue(new_obj);
-                nptr |= !gc_old(o->header);
-                if (!gc_try_setmark_tag(o, GC_MARKED)) new_obj = NULL;
-            }
+                assert(((uintptr_t)new_obj & 0x1) == 0);
+                if (push_into_queue) {
+                    __builtin_prefetch(jl_astaggedvalue(new_obj), 1, 3);
+                    __builtin_prefetch(&((jl_value_t**)new_obj)[7], 1, 3);
+                    __builtin_prefetch(&((jl_value_t**)new_obj)[16], 1, 3);
```

Comment:
128 bytes after the header seems a bit unlikely to need? Is the assumption that gets the full cache line around it?